webclient/pages/servers.py

```diff
@@ -0,0 +1,236 @@
+    year = 400 * (date // (365 * 400 + 97))
+    remainder = date % (365 * 400 + 97)
+
+    if remainder >= 365 * 100 + 25:
+        # There are 25 leap years in the first 100 years after every 400th
+        # year, as every 400th year is a leap year.
+        year += 100
+        remainder -= 365 * 100 + 25
+
+        # There are 24 leap years in the next couple of 100 years.
+        year += 100 * (remainder // (365 * 100 + 24))
+        remainder = remainder % (365 * 100 + 24)
+
+    if remainder >= 365 * 4 and not is_leap_year(year):
+        # The first 4 year of the century are not always a leap year.
+        year += 4
+        remainder -= 365 * 4
+
+    # There is 1 leap year every 4 years.
+    year += 4 * (remainder // (365 * 4 + 1))
+    remainder = remainder % (365 * 4 + 1)
+
+    # The last (max 3) years to account for; the first one can be, but is not
+    # necessarily a leap year.
+    while remainder >= (366 if is_leap_year(year) else 365):
+        remainder -= 366 if is_leap_year(year) else 365
+        year += 1
+
+    # Skip the 29th of February in non-leap years.
+    if not is_leap_year(year) and remainder >= 31 + 29 - 1:
+        remainder += 1
+
+    for month, days in DAYS_IN_MONTH.items():
+        if remainder < days:
+            break
+        remainder -= days
+    day = remainder + 1
+
+    return f"{year}-{month:02d}-{day:02d}"
+
+
+def _fix_server_info(server):
+    # Set the server_id to the one based on the IPv4 or, if not available, to
+    # the IPv6 variant.
+    server["server_id"] = getattr(server, "ipv4", server["ipv6"])["server_id"]
+
+    # Make dates human readable.
+    server["info"]["start_date"] = _date_to_string(server["info"]["start_date"])
+    server["info"]["game_date"] = _date_to_string(server["info"]["game_date"])
+    server["time_first_seen"] = (
+        datetime.utcfromtimestamp(server["time_first_seen"]).strftime("%Y-%m-%d %H:%M:%S") + " UTC"
+    )
+    server["time_last_seen"] = (
+        datetime.utcfromtimestamp(server["time_last_seen"]).strftime("%Y-%m-%d %H:%M:%S") + " UTC"
+    )
+
+
+def _split_version(version):
+    """
+    For official versions we control the format; for patchpacks we do not. So
+    we are guestimating a bit here what is most likely happening. If it fails,
+    we just fall back to sorting by string.
+    """
+
+    if "-" in version:
+        version, _, extra = version.partition("-")
+    else:
+        extra = "stable"
+
+    # Check if this is most likely an official release and sort it correctly.
+    version_parts = version.split(".")
+    if len(version_parts) == 3:
+        return [int(p) for p in version_parts] + [extra]
+
+    # Check if this is a patchpack like jgrpp-0.31.0.
+    version_parts = extra.split(".")
+    if len(version_parts) == 3:
+        return [int(p) for p in version_parts] + [version]
+
+    # We did not recognize what this is, so just fall back to string comparison.
+    return [0, 0, 0, version]
```

Comment:
This drops the "extra" part.
I guess the initial partition should not discard the raw input.
`part1, _ part2 = version_string.partition("-")`